plugins/kubernetes/test/controllers/kubernetes_dashboard_controller_test.rb

```diff
@@ -36,30 +44,112 @@
-      data.each { |item| delete_ids_from item }
+  def pod_list(cluster_deploy_group)
+    empty_list_template.tap do |list|
+      release.release_docs.each do |release_doc|
+        list.items << pod_item(cluster_deploy_group, release_doc) if release_doc.deploy_group == cluster_deploy_group.deploy_group
+      end
     end
   end
+
+  def pod_item(cluster_deploy_group, release_doc)
+    pod_list_item(release_doc.kubernetes_release.project, release_doc.kubernetes_release, release_doc.kubernetes_role, cluster_deploy_group)
+  end
+
+  def empty_list_template
+    RecursiveOpenStruct.new({
+      'kind': 'PodList',
+      'apiVersion': 'v1',
+      'items': []
+    })
+  end
+
+  def pod_list_item(project, release, role, cluster_deploy_group)
+    {
+      'metadata': {
+        'name': 'pod-name',
+        'namespace': cluster_deploy_group.namespace,
+        'labels': {
+          'project_id': project.id,
+          'release_id': release.id,
+          'deploy_group_id': cluster_deploy_group.deploy_group.id,
+          'role_id': role.id
+        }
+      },
+      'status': {
+        'phase': 'Running',
+        'conditions': [
+          {
+            'type': 'Ready',
+            'status': 'True'
+          }
+        ]
+      }
+    }
+  end
+
+  def expected_response
+    [].tap { |response|
+      release.release_docs.each { |release_doc|
+        add_role(response, release_doc).tap { |role|
+          add_deploy_group(role, release_doc).tap { |deploy_group|
+            add_release(deploy_group, release_doc)
+          }
+        }
+      }
+    }
+  end
+
+  def add_role(response, release_doc)
+    response << role_template(release_doc.kubernetes_role) unless response.any? { |role|
+      role.id == release_doc.kubernetes_role.id
+    }
+
+    response.find { |role| role.id == release_doc.kubernetes_role.id }
+  end
+
+  def add_deploy_group(role, release_doc)
+    role.deploy_groups << deploy_group_template(release_doc.deploy_group) unless role.deploy_groups.any? { |deploy_group|
+      deploy_group.id == release_doc.deploy_group.id
+    }
+
+    role.deploy_groups.find { |deploy_group| deploy_group.id == release_doc.deploy_group.id }
+  end
+
+  def add_release(deploy_group, release_doc)
+    deploy_group.releases << release_template(release, release_doc) unless deploy_group.releases.any? { |release|
+      release.id == release_doc.kubernetes_release.id
+    }
+  end
+
+  def role_template(role)
+    RecursiveOpenStruct.new({
+      'id': role.id,
+      'name': role.name,
+      'deploy_groups': []
+    })
+  end
+
+  def deploy_group_template(deploy_group)
+    RecursiveOpenStruct.new({
+      'id': deploy_group.id,
+      'name': deploy_group.name,
+      'releases': []
+    })
+  end
+
+  def release_template(release, release_doc)
+    RecursiveOpenStruct.new({
+      'id': release.id,
+      'build': release.build.label,
+      'target_replicas': release_doc.replica_target,
+      'live_replicas': 1
+    })
+  end
```

Comment:
I had to change it because I wanted to generate the whole response payload programatically, from all the `releases` and `release_docs` coming from fixtures. The alternative would have been to access each item in the response items array and update manually each of the values that I wanted to make sure that were added to the response. It wouldn't be very flexible.

Regarding your (valid) concerns, any error in the test code that generates the expected responses should be caught right away as the production code, in the controller, is tightly coupled to the structure of the responses coming from the Cluster. Finally, the response from the controller is validated in the test (`expected_json.must_equal actual_json`) so, any error should be easily detected. I find it really hard to get a false positive in this case.